### ISSUE TITLE
Jetty Client 9.4.x incorrectly handles too large fields from nginx 1.14 server

### ISSUE THREAD

**Issue by cupendra** [eclipse/jetty.project#2571]
The below code tests whether the target supports http/2 or not. This is part of our application code base, where we test if http/2 is supported on server, if not fall back to http/1.

This program use to work until the target server is with nginx 1.12, but when we migrate the server to nginx 1.14 this stopped working and always report java.util.concurrent.TimeoutException at line

  phaser.awaitAdvanceInterruptibly(phaser.arrive(), 5, TimeUnit.SECONDS);
But, when we made the same request through browsers(FireFox, Chrome) it works normal.

And the same program when tried with a different host like "webtide.com", the program works normal.

So, not sure where exactly is the problem? Whether the server has a problem as it works with browsers or the client has problem as it works normal with webtide.com and previous version of nginx.

Need help to resolve the issue on client side so that we can keep our nginx migration.

jetty http client version is 9.4.x, trying with jdk 10.0.1

``` java
public class Http2Test {

        private static boolean status = false;

        public static boolean isHttp2Supported(String host, int port) {
            long startTime = System.nanoTime();
            HTTP2Client client = null;
            try {
                System.out.println("Trying Http2Test with host::" + host + ", port::" + port);
                client = new HTTP2Client();
                SslContextFactory sslContextFactory = new SslContextFactory();
                client.addBean(sslContextFactory);
                client.start();

                FuturePromise<Session> sessionPromise = new FuturePromise<>();
                client.connect(sslContextFactory, new InetSocketAddress(host, port), new ServerSessionListener.Adapter(), sessionPromise);
                Session session = sessionPromise.get(5, TimeUnit.SECONDS);

                String jettyVersion = Jetty.VERSION;
                String clientClassName = client.getClass().getName();
                String userAgenet = clientClassName + "/" + jettyVersion;

                //System.out.println("UserAgent::"+userAgenet);
                HttpFields requestFields = new HttpFields();
                requestFields.put("User-Agent", userAgenet);
                HttpURI uri = new HttpURI("https://" + host + ":" + port);
                MetaData.Request metaData = new MetaData.Request("GET", uri, HttpVersion.HTTP_2, requestFields);
                HeadersFrame headersFrame = new HeadersFrame(metaData, null, true);

                final Phaser phaser = new Phaser(2);
                session.newStream(headersFrame, new Promise.Adapter<>(), new Stream.Listener.Adapter() {
                    @Override
                    public void onHeaders(Stream stream, HeadersFrame frame) {
                        System.err.println(frame.getMetaData().getHttpVersion());

                        int version = frame.getMetaData().getHttpVersion().getVersion();
                        if (version == 20) {
                            status = true;
                        }

                        //System.out.println(version);
                        if (frame.isEndStream()) {
                            phaser.arrive();
                        }
                    }

                    @Override
                    public void onData(Stream stream, DataFrame frame, Callback callback) {
                        System.err.println(frame);
                        callback.succeeded();
                        if (frame.isEndStream()) {
                            phaser.arrive();
                        }
                    }

                    @Override
                    public Stream.Listener onPush(Stream stream, PushPromiseFrame frame) {
                        System.err.println(frame);
                        phaser.register();
                        return this;
                    }
                });

                phaser.awaitAdvanceInterruptibly(phaser.arrive(), 5, TimeUnit.SECONDS);
            } catch (Exception ex) {
                ex.printStackTrace();
                status = false;
            } finally {
                if (client != null) {
                    try {
                        client.stop();
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }

            return status;
        }

        public static void main(String args[]) {
            //boolean supported = isHttp2Supported("webtide.com", 443);
            boolean supported = isHttp2Supported("testhttp2.yp4y.net", 443);
            System.out.println("Is Supported : " + supported);
        }

    }
```

**Comment by joakime** [comment]
Can you detail the nginx configuration? so we can attempt to replicate this issue.

**Comment by cupendra** [comment]
Hi,

Please find the attached configuration files 
[default.conf.txt](https://github.com/eclipse/jetty.project/files/2035964/default.conf.txt)
[nginx.conf.txt](https://github.com/eclipse/jetty.project/files/2035966/nginx.conf.txt)

nginx.conf file and default.conf file. Just renamed them with extension .txt as git doesn't allow .conf file to be attached

**Comment by sbordet** [comment]
@cupendra please also attach the DEBUG logs of Jetty's HTTP2Client, they are normally key to understand why the server does not reply.

**Comment by sbordet** [comment]
Asked on StackOverflow: https://stackoverflow.com/questions/50511149/jetty-client9-4-x-communicating-with-nginx-server-over-http-2-fails-to-get-the

**Comment by cupendra** [comment]
@sbordet, please find attached the jetty DEBUG output file below. 
[debugoutput.log](https://github.com/eclipse/jetty.project/files/2047076/debugoutput.log)

**Comment by sbordet** [comment]
@cupendra thanks. From the logs it is evident that there are other errors such as:

```
org.eclipse.jetty.http.BadMessageException: 431: Indexed field value too large
```

I'll analyze the log in details, meanwhile can you tell whether you are sending back huge cookies to the client ?

**Comment by cupendra** [comment]
@sbordet, thanks for looking into this. there are no cookies. I send this request through Firefox and I don't see any cookies sent for this request

**Comment by sbordet** [comment]
@gregw can you look at this issue ?

**Comment by sbordet** [comment]
@gregw:

```java
    @Test
    public void testResize() throws Exception
    {
        String encoded = "208c7689Aa6355E580Ae1697076196Df697e9413Ea681fA50400BcA01dB800Dc6595";
        ByteBuffer buffer = ByteBuffer.wrap(TypeUtil.fromHexString(encoded));
        HpackDecoder decoder = new HpackDecoder(4096, 8192);
        MetaData metaData = decoder.decode(buffer);
        System.err.println("metaData = " + metaData);
    }
```

**Comment by gregw** [comment]
It appears that nginx is sending bytes that sets the hpack dynamic table to 0, and then sending an encoded field that it wants to be indexed.  But the dynamic table size is 0, so we don't index it.   The question is, are we correct treating this as an error or should we just continuing??  I'll have to re-read the RFC! stand by....

**Comment by gregw** [comment]
So [RFC7541 4.4](https://tools.ietf.org/html/rfc7541#section-4.4) says:

> It is not an error to attempt to add an entry that is larger than the maximum size; an attempt to add an entry larger than the maximum size causes the table  to be emptied of all existing entries and results in an empty table.

So it is unclear why we [treat this as a BadMessage](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackDecoder.java#L250)

**Comment by cupendra** [comment]
@gregw, @sbordet any update on this? Do you need any more information from our nginx 1.14 server? if not, we would like to rollback our server to older version.

**Comment by joakime** [comment]
@cupendra the fix has been committed to `jetty-9.4.x` branch.

There are 3 choices for you:

1. Build the branch (`jetty-9.4.x`) and test it yourself. (best choice)
2. Wait till the overnight nightly snapshot build pushes the latest changes to the Jetty Snapshots repository at https://oss.sonatype.org/content/repositories/jetty-snapshots/ and then use the latest SNAPSHOT build to test it.
3. Wait until the next formal release and then test it.

**Comment by cupendra** [comment]
Okay. Thanks @joakime
